File: src/mod10_design_pattern/lesson10_flyweight/java/CreatureFactory.java
package mod10_design_pattern.lesson10_flyweight.java;

import java.util.HashMap;
import java.util.Map;

public class CreatureFactory {

    private Map<String, MagicalCreature> creatureMap = new HashMap<>();

    public MagicalCreature getCreature(String type, String color){

        String key= type+":"+color;//"owl:white"

        if (!creatureMap.containsKey(key)){
            switch(type){
                case "owl":
                    creatureMap.put(key,new Owl(type, color));
                    break;
                case "cat":
                    creatureMap.put(key,new Cat(type, color));
                    break;
                default:
                    throw new IllegalArgumentException("Unknown creature type");
            }
        }
        return creatureMap.get(key);
    }
}

File: src/mod10_design_pattern/lesson10_flyweight/kotlin/CreatureFactory.kt
package mod10_design_pattern.lesson10_flyweight.kotlin

class CreatureFactory {
    private val cratures = mutableMapOf<String, MagicalCreature>()

    fun getCreature(type: String, color: String): MagicalCreature {
        val key = "$type:$color"
        return cratures.getOrPut(key) {
            when (type) {
                "owl" -> Owl(type, color)
                "cat" -> Cat(type, color)
                else -> throw IllegalArgumentException("Unknown creature type")
            }
        }
    }
}